moshi/src/main/java/com/squareup/moshi/JsonWriter.kt

```diff
@@ -0,0 +1,553 @@
+
+  /**
+   * A string containing a full set of spaces for a single level of indentation, or null for no
+   * pretty printing.
+   */
+  @JvmField
+  internal var indent: String? = null
+
+  /**
+   * Configure this writer to relax its syntax rules. By default, this writer only emits well-formed
+   * JSON as specified by [RFC 7159](http://www.ietf.org/rfc/rfc7159.txt). Setting the
+   * writer to lenient permits the following:
+   *
+   * - Top-level values of any type. With strict writing, the top-level value must be an object
+   * or an array.
+   * - Numbers may be [NaNs][Double.isNaN] or [infinities][Double.isInfinite].
+   *
+   *  Returns true if this writer has relaxed syntax rules.
+   */
+  @get:CheckReturnValue
+  public var isLenient: Boolean = false
+
+  /**
+   * Sets whether object members are serialized when their value is null. This has no impact on
+   * array elements. The default is false.
+   *
+   * Returns true if object members are serialized when their value is null. This has no impact on
+   * array elements. The default is false.
+   */
+  @get:CheckReturnValue
+  public var serializeNulls: Boolean = false
+
+  @JvmField
+  internal var promoteValueToName: Boolean = false
+
+  /**
+   * Controls the deepest stack size that has begin/end pairs flattened:
+   *
+   * - If -1, no begin/end pairs are being suppressed.
+   * - If positive, this is the deepest stack size whose begin/end pairs are eligible to be flattened.
+   * - If negative, it is the bitwise inverse (~) of the deepest stack size whose begin/end pairs have been flattened.
+   *
+   * We differentiate between what layer would be flattened (positive) from what layer is being
+   * flattened (negative) so that we don't double-flatten.
+   *
+   * To accommodate nested flattening we require callers to track the previous state when they
+   * provide a new state. The previous state is returned from [beginFlatten] and restored
+   * with [endFlatten].
+   */
+  @JvmField
+  internal var flattenStackSize: Int = -1
+
+  private var tags: MutableMap<Class<*>, Any>? = null
+
+  /** Returns the scope on the top of the stack.  */
+  internal fun peekScope(): Int {
+    check(stackSize != 0) { "JsonWriter is closed." }
+    return scopes[stackSize - 1]
+  }
+
+  /** Before pushing a value on the stack this confirms that the stack has capacity.  */
+  internal fun checkStack(): Boolean {
+    if (stackSize != scopes.size) return false
+    if (stackSize == 256) {
+      throw JsonDataException("Nesting too deep at $path: circular reference?")
+    }
+    scopes = scopes.copyOf(scopes.size * 2)
+    pathNames = pathNames.copyOf(pathNames.size * 2)
+    pathIndices = pathIndices.copyOf(pathIndices.size * 2)
+    if (this is JsonValueWriter) {
+      stack = stack.copyOf(stack.size * 2)
+    }
+    return true
+  }
+
+  internal fun pushScope(newTop: Int) {
+    scopes[stackSize++] = newTop
+  }
+
+  /** Replace the value on the top of the stack with the given value.  */
```

Comment:
```suggestion
  /** Replace the value on the top of the stack with the given value. */
```

Please clean up all of these single line docs with a double space at the end